src/src.cpp

```diff
@@ -13,9 +13,58 @@
 #endif
 #include <oomph/util/heap_pimpl.hpp>
 #include <oomph/util/stack_pimpl.hpp>
+#include <oomph/barrier.hpp>
+#include "./thread_id.hpp"
+#include <map>
+#include <set>
+#include <mutex>
 
 namespace oomph
 {
+///////////////////////////////
+// thread_id                 //
+///////////////////////////////
+namespace
+{
+std::uintptr_t*
+alloc_tid_m()
+{
+    auto ptr = new std::uintptr_t{};
+    *ptr = (std::uintptr_t)ptr;
+    return ptr;
+}
+} // namespace
+
+thread_id::thread_id()
+: m{alloc_tid_m()}
+{
+}
+
+thread_id::~thread_id() { delete m; }
+
+thread_id const&
+tid()
+{
+    static thread_local thread_id id;
+    return id;
+}
+
+namespace
+{
+std::mutex comm_map_mtx;
+std::map<context_impl const*,
+    std::map<std::uintptr_t, std::set<std::pair<communicator_impl*, communicator::schedule*>>>>
```

Comment:
why do you keep both the implementation, and the schedule in the map? do you use it? seems to me that `communicator_impl*` would be enough?

Reply:
also, I don't understand why is this a `set::set`. Is it possible to attempt to insert a duplicate pair `<communicator_impl*, communicator::schedule*>`? You do it in the constructor below. I guess `communicator_impl*` could be passed multiple times, but `m_schedule{std::make_unique<schedule>()}` is always unique.

Reply:
I think I can remove the schedule for now, seems it is not needed at this point.
I could use a `std::vector` instead of a `std::set`, but we need to erase and insert whenever a communicator is destroyed or created, respectively, so a set makes more sense